plugins/gh-pages.lisp

```diff
@@ -1,33 +1,164 @@
+
 
 (defpackage :coleslaw-gh-pages
   (:use :cl)
-  (:import-from :puri #:parse-uri #:uri-host)
   (:import-from :coleslaw #:*config*
-                          #:deploy
-                          #:deploy-dir
+                          #:rel-path
+                          #:fmt
+                          #:run-program
+                          #:staging-dir
                           #:domain
-                          #:rel-path)
+                          #:with-current-directory
+                          #:update-symlink)
   (:export #:enable))
 
 (in-package :coleslaw-gh-pages)
 
+
+(defvar *datetime-format*
+  '((:YEAR 4) #\- (:MONTH 2) #\- (:DAY 2)
+    #\Space
+    (:HOUR 2) #\: (:MIN 2))
+  "Date and time format for commit message.")
+
+
 (defvar *cname* nil
   "The domain CNAME for github to serve pages from.")
 
-(defmethod deploy :after (staging)
-  (let ((blog (truename (rel-path (deploy-dir *config*) ".curr"))))
-    (delete-file (rel-path blog "index.html"))
-    (cl-fad:copy-file (rel-path blog "1.html") (rel-path blog "index.html"))
-    (with-open-file (out (rel-path blog "CNAME")
-                     :direction :output
-                     :if-exists :supersede
-                     :if-does-not-exist :create)
-      (format out "~A~%" *cname*))))
-
-(defun enable (&key cname)
-  (typecase cname
-    (string (setf *cname* cname))
-    (t (setf *cname* (uri-host (parse-uri (domain *config*)))))
-    (otherwise (error "Not a valid CNAME: ~A" cname))))
+
+(defvar *index-page* nil
+  "Name of the page to make an \"index\". For example: \"about.html\".")
+
+(defvar *push-to-github* nil
+  "If true, then new site will be pushed to the github after successful deployment.")
+
+
+(defun run (program &rest args)
+  (inferior-shell:run (fmt program args)
+                      :show t
+                      :output :interactive
+                      :error-output :interactive))
+
+
+(defun empty-string-p (string)
+  (eql (length string)
+       0))
+
+
+(defun get-git-branches (&key (remote nil))
+  "Returns remote branches for current git repository"
+  (let* ((output (with-output-to-string (s)
+                   (inferior-shell:run (if remote
+                                           "git branch --remote"
+                                           "git branch")
+                                       :output s)))
+         (lines (remove-if #'empty-string-p
+                           (split-sequence:split-sequence #\Newline output)))
+         (stripped-lines (mapcar (lambda (line)
+                                   (subseq line 2))
+                                 lines)))
+    stripped-lines))
+
+
+(defun switch-to-branch (repo-dir branch)
+  "Switches git repository to a given branch or creates a new branch."
+  (with-current-directory repo-dir
+    (let ((remote-branches (get-git-branches :remote t))
+          (local-branches (get-git-branches))
+          (remote-branch (concatenate 'string "origin/" branch)))
```

Comment:
@Ferada Should it be moved the configuration variable, ej `*remote-name*` or do you have another solution in mind